### PR TITLE
UCT/BASE: uct_mem_alloc reinit addr/length before using any method

### PR DESCRIPTION
**Pull request by Akshay-Venkatesh**
## What/ Why ?
`alloc_length` and `address` variables in uct_mem_alloc may not retain user requested values after one or methods are attempted. This patch re-initializes the variables to the original user requested values.

### CODE REVIEW

**Comment by Akshay-Venkatesh** [comment]
cc @yosefe @tonycurtis

File: src/uct/base/uct_mem.c [hunk 1]
```diff
@@ -215,6 +218,9 @@ ucs_status_t uct_mem_alloc(size_t length, const uct_alloc_method_t *methods,
                 break;
             }
 
+            alloc_length = length;
+            address      = (params->field_mask & UCT_MEM_ALLOC_PARAM_FIELD_ADDRESS) ?
+                           params->address : NULL;
```
Comment (yosefe) [inline src/uct/base/uct_mem.c:223]: we an just use "length" for this (since posix_memalign does not change it)

File: src/uct/base/uct_mem.c [hunk 2]
```diff
@@ -249,8 +258,9 @@ ucs_status_t uct_mem_alloc(size_t length, const uct_alloc_method_t *methods,
 
 #ifdef SHM_HUGETLB
             /* Allocate huge pages */
-            address = (flags & UCT_MD_MEM_FLAG_FIXED) ?
-                      params->address : NULL;
+            alloc_length = length;
+            address      = (flags & UCT_MD_MEM_FLAG_FIXED) ?
```
Comment (yosefe) [inline src/uct/base/uct_mem.c:262]: UCT_MEM_ALLOC_PARAM_FIELD_ADDRESS ?

File: src/uct/base/uct_mem.c [hunk 3]
```diff
@@ -231,6 +237,9 @@ ucs_status_t uct_mem_alloc(size_t length, const uct_alloc_method_t *methods,
 
             /* Request memory from operating system using mmap() */
 
+            alloc_length = length;
+            address      = (params->field_mask & UCT_MEM_ALLOC_PARAM_FIELD_ADDRESS) ?
+                           params->address : NULL;
```
Comment (yosefe) [inline src/uct/base/uct_mem.c:242]: maybe do this in the beginning of 'for' loop?

**Comment by yosefe** [comment]
ok to test

File: src/uct/base/uct_mem.c (hunk 1 above)
Comment (Akshay-Venkatesh) [reply src/uct/base/uct_mem.c:223]: but `alloc_length` could be changed by uct_md_mem_alloc call but md_mem_alloc failed, right? unless failure in md_mem_alloc is a fatal error and no other methods are attempted.

Comment (Akshay-Venkatesh) [reply src/uct/base/uct_mem.c:223]: It does turn out to be a non-recoverable error. I'll remove this.

File: src/uct/base/uct_mem.c (hunk 2 above)
Comment (Akshay-Venkatesh) [reply src/uct/base/uct_mem.c:262]: Haven't really change the logic here. This is how it looks in 1.9.x
```
#ifdef SHM_HUGETLB
            /* Allocate huge pages */
            alloc_length = min_length;
            address = (flags & UCT_MD_MEM_FLAG_FIXED) ? addr : NULL;
            status = ucs_sysv_alloc(&alloc_length, min_length * 2, &address,
                                    SHM_HUGETLB, alloc_name, &shmid);
            if (status == UCS_OK) {
                goto allocated_without_md;
            }
#else
```

Seems like sysv_alloc requires FLAG_FIXED to be passed. Should we change the logic here?

Comment (yosefe) [reply src/uct/base/uct_mem.c:261]: ok

File: src/uct/base/uct_mem.c [hunk 4]
```diff
@@ -215,8 +218,10 @@ ucs_status_t uct_mem_alloc(size_t length, const uct_alloc_method_t *methods,
                 break;
             }
 
+            address      = (params->field_mask & UCT_MEM_ALLOC_PARAM_FIELD_ADDRESS) ?
+                           params->address : NULL;
```
Comment (yosefe) [inline src/uct/base/uct_mem.c:222]: maybe add inline function uct_mem_alloc_params_get_address(params) -> address (or NULL)

**Comment by shamisp** [comment]
@Akshay-Venkatesh this does not look like API change ? The title is a bit confusing.

File: src/uct/base/uct_mem.c (hunk 4 above)
Comment (Akshay-Venkatesh) [reply src/uct/base/uct_mem.c]: done

**Review by yosefe — APPROVED** [review]

**Pull request merged**